File: [23.12]Design-Pattern/Java/src/behavioral/interpreter/Division.java
package behavioral.interpreter;

public class Division implements Expression {
    private Expression leftOperand;
    private Expression rightOperand;

    public Division(Expression leftOperand, Expression rightOperand) {
        this.leftOperand = leftOperand;
        this.rightOperand = rightOperand;
    }

    @Override
    public double interpret() {
        if (rightOperand.interpret() == 0) {
            throw new ArithmeticException("Division by zero");
        }
        return leftOperand.interpret() / rightOperand.interpret();
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/interpreter/Division.kt
package behavioral.interpreter

class Division(private val leftExpression: Expression, private val rightExpression: Expression): Expression {

    override fun interpret(): Double {
        if (rightExpression.interpret() == 0.0) throw ArithmeticException("Division by zero")
        return leftExpression.interpret() / rightExpression.interpret()
    }
}